File: packages/node-sdk/src/http-client.ts
import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { createAuthHeaders, createGetAuthHeaders } from './auth';

/**
 * HTTP client for making authenticated requests to the Wusul API
 */
export class HttpClient {
  private client: AxiosInstance;
  private accountId: string;
  private sharedSecret: string;

  constructor(accountId: string, sharedSecret: string, baseUrl: string, timeout: number) {
    this.accountId = accountId;
    this.sharedSecret = sharedSecret;

    this.client = axios.create({
      baseURL: baseUrl,
      timeout,
      headers: {
        'Content-Type': 'application/json',
      },
    });

    // Add response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => {
        if (error.response) {
          // The request was made and the server responded with a status code
          // that falls out of the range of 2xx
          const errorMessage = error.response.data?.error || error.response.data?.message || error.message;
          throw new Error(`Wusul API Error (${error.response.status}): ${errorMessage}`);
        } else if (error.request) {
          // The request was made but no response was received
          throw new Error('No response received from Wusul API');
        } else {
          // Something happened in setting up the request that triggered an Error
          throw new Error(`Request setup error: ${error.message}`);
        }
      }
    );
  }

  /**
   * Makes a GET request
   */
  async get<T>(url: string, sigPayload?: Record<string, any>): Promise<T> {
    const { headers, sigPayload: encodedPayload } = createGetAuthHeaders(
      this.accountId,
      this.sharedSecret,
      sigPayload
    );

    const response: AxiosResponse<any> = await this.client.get(url, {
      headers,
      params: {
        sig_payload: encodedPayload,
      },
    });

    return response.data.data || response.data;
  }

  /**
   * Makes a POST request
   */
  async post<T>(url: string, data?: any): Promise<T> {
    const headers = createAuthHeaders(this.accountId, this.sharedSecret, data);
    const response: AxiosResponse<any> = await this.client.post(url, data, { headers });
    return response.data.data || response.data;
  }

  /**
   * Makes a PATCH request
   */
  async patch<T>(url: string, data?: any): Promise<T> {
    const headers = createAuthHeaders(this.accountId, this.sharedSecret, data);
    const response: AxiosResponse<any> = await this.client.patch(url, data, { headers });
    return response.data.data || response.data;
  }

  /**
   * Makes a DELETE request
   */
  async delete<T>(url: string): Promise<T> {
    const headers = createAuthHeaders(this.accountId, this.sharedSecret);
    const response: AxiosResponse<any> = await this.client.delete(url, { headers });
    return response.data.data || response.data;
  }
}

File: packages/java-sdk/src/main/java/io/wusul/sdk/http/HttpClient.java
package io.wusul.sdk.http;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.wusul.sdk.auth.Auth;
import okhttp3.*;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * HTTP client for making authenticated requests to the Wusul API.
 */
public class HttpClient {
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final ObjectMapper objectMapper = new ObjectMapper()
            .registerModule(new JavaTimeModule());

    private final OkHttpClient client;
    private final String accountId;
    private final String sharedSecret;
    private final String baseUrl;

    /**
     * Creates a new HttpClient instance.
     *
     * @param accountId The account ID
     * @param sharedSecret The shared secret
     * @param baseUrl The base URL for the API
     * @param timeout The request timeout in milliseconds
     */
    public HttpClient(String accountId, String sharedSecret, String baseUrl, long timeout) {
        this.accountId = accountId;
        this.sharedSecret = sharedSecret;
        this.baseUrl = baseUrl;

        this.client = new OkHttpClient.Builder()
                .connectTimeout(timeout, TimeUnit.MILLISECONDS)
                .readTimeout(timeout, TimeUnit.MILLISECONDS)
                .writeTimeout(timeout, TimeUnit.MILLISECONDS)
                .build();
    }

    /**
     * Makes a GET request.
     *
     * @param url The URL path
     * @param sigPayload Optional signature payload
     * @param responseType The response type class
     * @return The response object
     */
    public <T> T get(String url, Map<String, Object> sigPayload, Class<T> responseType) throws IOException {
        Auth.AuthHeadersResult authHeaders = Auth.createGetAuthHeaders(accountId, sharedSecret, sigPayload);

        HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + url).newBuilder();
        urlBuilder.addQueryParameter("sig_payload", authHeaders.getSigPayload());

        Request.Builder requestBuilder = new Request.Builder()
                .url(urlBuilder.build())
                .get();

        for (Map.Entry<String, String> header : authHeaders.getHeaders().entrySet()) {
            requestBuilder.addHeader(header.getKey(), header.getValue());
        }

        return executeRequest(requestBuilder.build(), responseType);
    }

    /**
     * Makes a POST request.
     *
     * @param url The URL path
     * @param payload The request payload
     * @param responseType The response type class
     * @return The response object
     */
    public <T> T post(String url, Object payload, Class<T> responseType) throws IOException {
        Map<String, String> headers = Auth.createAuthHeaders(accountId, sharedSecret, payload);

        RequestBody body = payload != null
                ? RequestBody.create(objectMapper.writeValueAsString(payload), JSON)
                : RequestBody.create("", JSON);

        Request.Builder requestBuilder = new Request.Builder()
                .url(baseUrl + url)
                .post(body);

        for (Map.Entry<String, String> header : headers.entrySet()) {
            requestBuilder.addHeader(header.getKey(), header.getValue());
        }

        return executeRequest(requestBuilder.build(), responseType);
    }

    /**
     * Makes a PATCH request.
     *
     * @param url The URL path
     * @param payload The request payload
     * @param responseType The response type class
     * @return The response object
     */
    public <T> T patch(String url, Object payload, Class<T> responseType) throws IOException {
        Map<String, String> headers = Auth.createAuthHeaders(accountId, sharedSecret, payload);

        RequestBody body = RequestBody.create(objectMapper.writeValueAsString(payload), JSON);

        Request.Builder requestBuilder = new Request.Builder()
                .url(baseUrl + url)
                .patch(body);

        for (Map.Entry<String, String> header : headers.entrySet()) {
            requestBuilder.addHeader(header.getKey(), header.getValue());
        }

        return executeRequest(requestBuilder.build(), responseType);
    }

    /**
     * Makes a DELETE request.
     *
     * @param url The URL path
     * @param responseType The response type class
     * @return The response object
     */
    public <T> T delete(String url, Class<T> responseType) throws IOException {
        Map<String, String> headers = Auth.createAuthHeaders(accountId, sharedSecret, null);

        Request.Builder requestBuilder = new Request.Builder()
                .url(baseUrl + url)
                .delete();

        for (Map.Entry<String, String> header : headers.entrySet()) {
            requestBuilder.addHeader(header.getKey(), header.getValue());
        }

        return executeRequest(requestBuilder.build(), responseType);
    }

    /**
     * Executes a request and handles the response.
     *
     * @param request The HTTP request
     * @param responseType The response type class
     * @return The response object
     */
    private <T> T executeRequest(Request request, Class<T> responseType) throws IOException {
        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                String errorBody = response.body() != null ? response.body().string() : "";
                String errorMessage;

                try {
                    JsonNode errorNode = objectMapper.readTree(errorBody);
                    errorMessage = errorNode.has("error") ? errorNode.get("error").asText()
                            : errorNode.has("message") ? errorNode.get("message").asText()
                            : "Unknown error";
                } catch (Exception e) {
                    errorMessage = errorBody.isEmpty() ? "Unknown error" : errorBody;
                }

                throw new IOException("Wusul API Error (" + response.code() + "): " + errorMessage);
            }

            String responseBody = response.body().string();
            JsonNode jsonNode = objectMapper.readTree(responseBody);

            // Extract data field if present, otherwise use the entire response
            JsonNode dataNode = jsonNode.has("data") ? jsonNode.get("data") : jsonNode;

            return objectMapper.treeToValue(dataNode, responseType);
        }
    }
}
